test/sm2_internal_test.c

```diff
@@ -36,6 +37,7 @@ static int get_faked_bytes(unsigned char *buf, int num)
     if (fake_rand_bytes == NULL)
         return saved_rand->bytes(buf, num);
 
+    OPENSSL_assert(fake_rand_bytes_offset + num <= fake_rand_size);
```

Comment:
`TEST_size_t_le` seems more appropriate

Reply:
This is in the callback. The idea was to abort the test immediately, without need for address sanitizer.
Actually I can remove this, and rely on the TEST_size_t_eq at the end.
If address sanitizer is running it will print the far more informative message than the assert.

Reply:
Remove, okay?

Reply:
Personally, I'd rather "return 0" immediately rather than rely on sanitizer to catch the issue. It will enable us to set a breakpoint in the code and easily figure out which call is causing us to ask for too much data. As you say it will still be caught in later TEST_size_t_eq.

Reply:
Sorry I missed your last message...
It is most of the time very convenient to get the call stack printed out instead of having to use
the debugger.
But if you want, I can change that again.

Reply:
I would much prefer it, but won't insist.

Reply:
sure, if you like it better that way.
It will likely not hit the TEST_size_t eq in that case,
because the random generator will appear to have failed the operation.

Reply:
ok, done.

Reply:
Argh!!!

I tried to shorten the fake_rand_size, and...
```
diff --git a/test/sm2_internal_test.c b/test/sm2_internal_test.c
index d4ec01e..6aeea23 100644
--- a/test/sm2_internal_test.c
+++ b/test/sm2_internal_test.c
@@ -245,7 +245,7 @@ static int sm2_crypt_test(void)
             "encryption standard",
             "004C62EEFD6ECFC2B95B92FD6C3D9575148AFA17425546D49018E5388D49DD7B4F"
             "003da18008784352192d70f22c26c243174a447ba272fec64163dd4742bae8bc98"
-            "00df17605cf304e9dd1dfeb90c015e93b393a6f046792f790a6fa4228af67d9588",
+            "00df1760f304e9dd1dfeb90c015e93b393a6f046792f790a6fa4228af67d9588",
             "307B0220245C26FB68B1DDDDB12C4B6BF9F2B6D5FE60A383B0D18D1C4144ABF17F"
             "6252E7022076CB9264C2A7E88E52B19903FDC47378F605E36811F5C07423A24B84"
             "400F01B80420BE89139D07853100EFA763F60CBE30099EA3DF7F8F364F9D10A5E9"
```

Unfortunately if the TEST_size_t_eq fails, the test still passes:
```
    # Subtest: ../../test/sm2_internal_test
    1..2
    # ERROR: (size_t) 'fake_rand_bytes_offset + num <= fake_rand_size' failed @ 
test/sm2_internal_test.c:40
    # [99] compared to [98]
    # ERROR: (bool) 'sm2_encrypt(key, digest, (const uint8_t *)message, msg_len,
 ctext, &ctext_len) == true' failed @ test/sm2_internal_test.c:178
    # false
```

How can this be????